api/src/api/users.rs

```diff
@@ -0,0 +1,121 @@
+use crate::models::response;
+use actix_web::{HttpResponse, get, post, web};
+use futures::stream::StreamExt;
+use mongodb::{Collection, Database, bson::doc};
+use serde_json::to_value;
+
+#[derive(serde::Deserialize, Clone)]
+pub struct UserPost {
+    pub username: String,
+    pub email: String,
+    pub password: String,
+}
+
+#[post("/users")]
+pub async fn post_user(db: web::Data<Database>, body: web::Json<UserPost>) -> HttpResponse {
+    let body = body.clone();
+
+    // check that a user with this username/email does not already exist
+    // TODO: drop usernames altogether and simply use email?
+    let user_collection: Collection<mongodb::bson::Document> = db.collection("users");
+    let existing_user = user_collection
+        .find_one(doc! { "username": &body.username })
+        .await;
+    match existing_user {
+        Ok(Some(_)) => {
+            return HttpResponse::BadRequest().body("user with this username already exists");
+        }
+        Ok(None) => {}
+        Err(e) => {
+            return HttpResponse::InternalServerError().body(format!(
+                "failed to query database for existing user. error: {}",
+                e
+            ));
+        }
+    }
+    let existing_user = user_collection
+        .find_one(doc! { "email": &body.email })
+        .await;
+    match existing_user {
+        Ok(Some(_)) => {
+            return HttpResponse::BadRequest().body("user with this email already exists");
+        }
+        Ok(None) => {}
+        Err(e) => {
+            return HttpResponse::InternalServerError().body(format!(
+                "failed to query database for existing user. error: {}",
+                e
+            ));
+        }
+    }
+
+    // create a new user document
+    // first, hash password
+    // TODO: permissions?
```

Comment:
@Theodlz is there already a permissions model for this application, or is this not important at this time?